ddtrace/api.py

```diff
@@ -12,40 +13,89 @@
 
 TRACE_COUNT_HEADER = 'X-Datadog-Trace-Count'
 
+_VERSIONS = {'v0.4': {'traces': '/v0.4/traces',
+                      'services': '/v0.4/services',
+                      'compatibility_mode': False,
+                      'fallback': 'v0.3'},
+             'v0.3': {'traces': '/v0.3/traces',
+                     'services': '/v0.3/services',
+                      'compatibility_mode': False,
+                      'fallback': 'v0.2'},
+             'v0.2': {'traces': '/v0.2/traces',
+                     'services': '/v0.2/services',
+                      'compatibility_mode': True,
+                      'fallback': None}}
+
+def _parse_response_json(response):
+    """
+    Parse the content of a response object, and return the right type,
+    can be a string if the output was plain text, or a dictionnary if
+    the output was a JSON.
+    """
+    if hasattr(response, 'read'):
+        body = response.read()
+        try:
+            if not isinstance(body, str) and hasattr(body, 'decode'):
+                body = body.decode('utf-8')
+            if hasattr(body, 'startswith') and body.startswith('OK'):
+                # This typically happens when using a priority-sampling enabled
+                # library with an outdated agent. It still works, but priority sampling
+                # will probably send too many traces, so the next step is to upgrade agent.
+                log.debug("'OK' is not a valid JSON, please make sure trace-agent is up to date")
+                return
+            content = loads(body)
+            return content
+        except (ValueError, TypeError) as err:
+            log.debug("unable to load JSON '%s': %s" % (body, err))
+
 class API(object):
     """
     Send data to the trace agent using the HTTP protocol and JSON format
     """
-    def __init__(self, hostname, port, headers=None, encoder=None):
+    def __init__(self, hostname, port, headers=None, encoder=None, priority_sampling=False):
```

Comment:
I think if we let the caller of this method provide an API version instead of the `priority_sampling` flag, we would have a bit more flexibility (and we could also get rid of the conditional below).

Reply:
I think it's more a matter to simplify the caller logic, plus decoupling both components.

With this implementation, you [propagate](https://github.com/DataDog/dd-trace-py/pull/359/files#diff-c3dd6d04070c4b00a58e7ab4aeafd404R34) the `priority_sampling` boolean as a feature flag (available in the `.configure()` method), rather than requiring the `Writer` class to know which API version is the Priority Sampling API (that is known only by the `API` class).

Reply:
One more thing: all of this is internal machinery so not a public API people will play with.